src/driver-sinowealth.c

```diff
@@ -0,0 +1,475 @@
+
+	sinowealth_init_profile(device);
+
+	profile = ratbag_device_get_profile(device, 0);
+	rc = sinowealth_read_profile(profile);
+	if (rc) {
+		rc = -ENODEV;
+		goto err;
+	}
+
+	return 0;
+
+err:
+	free(drv_data);
+	ratbag_set_drv_data(device, 0);
+	return rc;
+}
+
+static int
+sinowealth_commit(struct ratbag_device *device)
+{
+	struct ratbag_profile *profile = ratbag_device_get_profile(device, 0);
+	struct sinowealth_data *drv_data = device->drv_data;
+	struct sinowealth_config_report *config = &drv_data->config;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	int rc;
+
+	config->config &= ~SINOWEALTH_XY_INDEPENDENT;
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (resolution->dpi_x != resolution->dpi_y && resolution->dpi_x && resolution->dpi_y) {
+			config->config &= SINOWEALTH_XY_INDEPENDENT;
+			break;
+		}
+	}
+
+	config->dpi_enabled = 0xFF;
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (config->config & SINOWEALTH_XY_INDEPENDENT) {
+			config->dpi[resolution->index * 2] = sinowealth_dpi_to_raw(resolution->dpi_x);
+			config->dpi[resolution->index * 2 + 1] = sinowealth_dpi_to_raw(resolution->dpi_y);
+		} else {
+			config->dpi[resolution->index] = sinowealth_dpi_to_raw(resolution->dpi_x);
+		}
+		if (resolution->dpi_x && resolution->dpi_y) {
+			/* enable DPI step (dpi_enabled is inverted) */
+			config->dpi_enabled &= ~(1<<resolution->index);
+		}
+	}
+
+	/* Body lighting */
+	led = ratbag_profile_get_led(profile, 0);
+	switch(led->mode) {
+	case RATBAG_LED_OFF:
+		config->rgb_effect = RGB_OFF;
+		break;
+	case RATBAG_LED_ON:
+		config->rgb_effect = RGB_SINGLE;
+		config->single_color = sinowealth_color_to_rbg(led->color);
+		break;
+	case RATBAG_LED_CYCLE:
+		config->rgb_effect = RGB_GLORIOUS;
+		config->glorious_mode = sinowealth_to_rgb_mode(led->brightness, led->ms);
+		break;
+	case RATBAG_LED_BREATHING:
+		config->rgb_effect = RGB_BREATHING1;
+		config->breathing1_color = sinowealth_color_to_rbg(led->color);
+		config->breathing1_mode = sinowealth_to_rgb_mode(led->brightness, led->ms);
+		break;
+	}
+	ratbag_led_unref(led);
+
+	/* DPI indicator LED */
+	for (int i = 1; i < SINOWEALTH_NUM_DPIS + 1; i++) {
+		led = ratbag_profile_get_led(profile, i);
+		config->dpi_color[i - 1] = sinowealth_color_to_raw(led->color);
+		ratbag_led_unref(led);
+	}
+
+	config->config_write = 0x7b; /* magic */
+
+	rc = ratbag_hidraw_set_feature_report(device, SINOWEALTH_REPORT_ID_CONFIG,
+					      (uint8_t*) config, SINOWEALTH_CONFIG_SIZE);
+	if(rc != SINOWEALTH_CONFIG_SIZE) {
+		log_error(device->ratbag, "Error while writing config: %d\n", rc);
+		ratbag_profile_unref(profile);
+		return -1;
+	}
+
+	ratbag_profile_unref(profile);
+	return 0;
+}
+
+static void
+sinowealth_remove(struct ratbag_device *device)
+{
+	ratbag_close_hidraw(device);
+	free(ratbag_get_drv_data(device));
+}
+
+struct ratbag_driver sinowealth_driver = {
+	.name = "Sinowealth Gaming Mouse",
```

Comment:
Maybe just leave it as `Sinowealth` to make it more consistent with the other drivers (usually it's the vendor / protocol name, only for very device specific drivers the name is in there).